karapace/kafka_admin.py

```diff
@@ -0,0 +1,290 @@
+                continue
+            else:
+                break
+        else:
+            raise AuthenticationFailedError()
+
+    def new_topic(
+        self,
+        name: str,
+        *,
+        num_partitions: int = 1,
+        replication_factor: int = 1,
+        config: dict[str, str] | None = None,
+        request_timeout: int = TOPIC_CREATION_TIMEOUT_S,
+    ) -> NewTopic:
+        new_topic = NewTopic(
+            topic=name,
+            num_partitions=num_partitions,
+            replication_factor=replication_factor,
+            config=config if config is not None else {},
+        )
+        LOG.info("Creating new topic %s with replication factor %s", new_topic, replication_factor)
+        futmap: dict[str, Future] = self.create_topics([new_topic], request_timeout=request_timeout)
+        try:
+            single_futmap_result(futmap)
+            return new_topic
+        except KafkaException as exc:
+            raise_from_kafkaexception(exc)
+
+    def update_topic_config(self, name: str, config: dict[str, str]) -> None:
+        LOG.info("Updating topic '%s' configuration with %s", name, config)
+        futmap = self.alter_configs([ConfigResource(ResourceType.TOPIC, name, set_config=config)])
+        try:
+            single_futmap_result(futmap)
+        except KafkaException as exc:
+            raise_from_kafkaexception(exc)
+
+    def delete_topic(self, name: str) -> None:
+        LOG.info("Deleting topic '%s'", name)
+        futmap = self.delete_topics([name])
+        try:
+            single_futmap_result(futmap)
+        except KafkaException as exc:
+            raise_from_kafkaexception(exc)
+
+    def cluster_metadata(self, topics: Iterable[str] | None = None) -> dict:
+        """Fetch cluster metadata and topic information for given topics or all topics if not given.
+
+        Using the `list_topics` method of the `AdminClient`, as this actually provides
+        metadata for the entire cluster, not just topics, as suggested by the name.
+
+        The topics filter is only applied _after_ fetching the cluster metadata,
+        due to `list_topics` only accepting a single topic as a filter.
+        """
+        LOG.info("Fetching cluster metadata with topic filter: %s", topics)
+        cluster_metadata: ClusterMetadata = self.list_topics()
+        topics_metadata: dict[str, TopicMetadata] = cluster_metadata.topics
+        brokers_metadata: dict[str, BrokerMetadata] = cluster_metadata.brokers
+
+        if topics is not None and any(topic not in topics_metadata.keys() for topic in topics):
+            raise UnknownTopicOrPartitionError()
+
+        topics_data: dict[str, dict] = {}
+        for topic, topic_metadata in topics_metadata.items():
+            if topics is not None and topic not in topics:
+                continue
+
+            partitions_data = []
+            for partition_id, partition_metadata in topic_metadata.partitions.items():
+                partition_data = {
```

Comment:
issue: This looks like a place where it's not really justified to use dicts, let's use dataclasses with proper typing here?

Reply:
For now I wanted to have the usage of the admin client change as little as possible (there are already quite a lot of changes here) and just have an almost-drop-in-replacement (as much as possible). If you are okay with it, I'll leave a TODO here and tackle this as a follow-up.